bindings/java/src/main/native/jucx_common_def.cc

```diff
@@ -232,104 +224,89 @@ static inline void jucx_call_callback(jobject callback, jobject jucx_request,
     }
 }
 
-UCS_PROFILE_FUNC_VOID(jucx_request_callback, (request, status), void *request, ucs_status_t status)
+UCS_PROFILE_FUNC_VOID(jucx_request_callback, (request, status, user_data), void *request,
+                      ucs_status_t status, void *user_data)
 {
-    struct jucx_context *ctx = (struct jucx_context *)request;
-    ucs_recursive_spin_lock(&ctx->lock);
-    if (ctx->jucx_request == NULL) {
-        // here because 1 of 2 reasons:
-        // 1. progress is in another thread and got here earlier then process_request happened.
-        // 2. this callback is inside ucp_tag_recv_nb function.
-        ctx->status = status;
-        ucs_recursive_spin_unlock(&ctx->lock);
-        return;
-    }
+    jobject jucx_request = reinterpret_cast<jobject>(user_data);
 
     JNIEnv *env = get_jni_env();
-    set_jucx_request_completed(env, ctx->jucx_request, ctx, status);
 
-    if (ctx->callback != NULL) {
-        jucx_call_callback(ctx->callback, ctx->jucx_request, status);
-        env->DeleteGlobalRef(ctx->callback);
+    set_jucx_request_completed(env, jucx_request, UCS_PTR_STATUS(status));
+    ucp_request_free(request);
+
+    jobject callback = env->GetObjectField(jucx_request, request_callback);
+
+    if (callback != NULL) {
+        jucx_call_callback(callback, jucx_request, status);
+        env->SetObjectField(jucx_request, request_callback, NULL);
```

Comment:
why need to set callback to NULL?

Reply:
There could be 1 callback object for many requests. I set everywhere to ULL, so requests objects could be garbage collected, and eventually callback as well, (to remove reference).